### PR TITLE
Env

### PR DESCRIPTION
**Pull request by stsp**
Checks for missing/empty environment.

### CODE REVIEW

**Comment by alexfru** [comment]
Want to introduce such checks in other places? There's related code in c0.c, getenv.c, system.c.

It's kinda odd that FreeDOS would have CR instead of 0 for an empty environment. OTOH, it should be a very rare situation for normal programs and normal DOS setups.

**Comment by stsp** [comment]
> Want to introduce such checks in other places? There's related code in c0.c, getenv.c,
> system.c.

No, it doesn't get there.
Stub fails to load an exe and exits, so
my patch just adds the better messages.

> It's kinda odd that FreeDOS would have CR instead of 0 for an empty environment.
> OTOH, it should be a very rare situation for normal programs and normal DOS setups.

Absolutely agree.
Plus they noticed that every shell
crash, and added a dummy `PATH=.`
as a work-around. So its not even
easy to replicate the problem w/o
undoing the work-around.
If you want, I can remove the freedos
case.

**Comment by stsp** [comment]
> Want to introduce such checks in other places? There's related code in c0.c, getenv.c,
> system.c.

No, it doesn't get there.
Stub fails to load an exe and exits, so
my patch just adds the better messages.

> It's kinda odd that FreeDOS would have CR instead of 0 for an empty environment.
> OTOH, it should be a very rare situation for normal programs and normal DOS setups.

Absolutely agree.
Plus they noticed that every shell
crash, and added a dummy `PATH=.`
as a work-around. So its not even
easy to replicate the problem w/o
undoing the work-around.
If you want, I can remove the freedos
case.

**Comment by stsp** [comment]
Removed freedos check.
It was for overall completeness, and
is not needed with the latest versions
of freedos.

**Comment by alexfru** [comment]
Yes, I'd prefer not to have the special/obscure FreeDOS case.

What I meant was that there are other places where the environment is manipulated in the library (startup code, getenv(), system(), tmpnam()), I mean, non-DPMI DOS code.

**Comment by stsp** [comment]
Ah, I guess you mean, we actually
can get there, just not in DPMI mode.
Yes, will take a look.

**Comment by stsp** [comment]
Do you want to pass the argv0
pointer from the stub, rather than
to always traverse the env strings?

Also it would really help if the
envp array is accessible from the
user's program (usually done via
an extra arg to main()), as currently
I also need to traverse the envs by
hands.

**Comment by stsp** [comment]
Added the checks to the places I
was able to find. Please let me know
where exactly to look for other such
instances. :)
This last patch is a bit untested, since
in DPMI mode we don't even get there.

**Comment by stsp** [comment]
> Do you want to pass the argv0
> pointer from the stub

You already do...
So the only remaining wish is to
have the access to envp in main().

**Comment by alexfru** [comment]
Ping.

**Comment by stsp** [comment]
What kind of ping?
I added this:
https://github.com/alexfru/SmallerC/pull/45/commits/3dc76acb8c5a1e6a0c11b62ae9f51e1bea09be3b

**Comment by alexfru** [comment]
In the stub you have this check:
~~~
        mov     es, [es:0x2c]   ; es = environment segment
        mov     [env_seg], es
        cmp     word [env_seg], 0
        je      err_noenv
~~~

In here
~~~
#ifndef _DPMI
static
int setargs(int* pargc, char*** pargv)
{
  unsigned psp = DosGetPspSeg();
  unsigned env = peek(psp, 0x2c);
  ...
  if (peekb(env, i) == 0) // initial 0 means no variables
  ...
~~~
you just use env without checking it for 0 first.

**Comment by stsp** [comment]
Added more checks.

**Comment by alexfru** [comment]
This introduces inconsistent behavior: DPMI programs will terminate if the environment segment is 0, whereas other DOS programs will carry on with under-constructed args. I'd prefer both terminate, e.g. you can call `exit(EXIT_FAILURE)` instead of returning 0 from setargs() and no need to check `env` for 0 in `getenv()`.

**Comment by stsp** [comment]
So like this?

**Comment by alexfru** [comment]
Could you add error printing as asked above?

**Comment by stsp** [comment]
Your message was edited about printing.
Also I wonder if stdio works properly in c0.c.
But I added now the msg with `puts()`.

**Comment by alexfru** [comment]
Can you see the message/review item, where I suggested to use __write()? That should work always, even if most of the library isn't initialized. Use that instead of puts().

**Comment by stsp** [comment]
No, sorry, I don't see that review.
Are you sure its not marked with
"Pending" flag?

Anyway, added `__write()` now.

File: v0100/srclib/c0.c [hunk 1]
```diff
@@ -305,10 +305,20 @@ int setargs(int* pargc, char*** pargv)
   char* params;
 
   // First, try to extract the full program name.
+  if (!env)
+  {
+    static char msg[] = "No environment segment!";
```
Comment (alexfru) [inline v0100/srclib/c0.c:310]: Include trailing \r\n in msg[].

**Comment by alexfru** [comment]
You're right, it was marked as pending and I must've "submitted" it. It's a bit confusing.

File: v0100/srclib/c0.c (hunk 1 above)
Comment (stsp) [reply v0100/srclib/c0.c]: I hate untested patches. :)

**Comment by stsp** [comment]
Done.

**Pull request merged**